code/modules/mob/living/silicon/robot/sprites/_sprite_datum.dm

```diff
@@ -27,7 +28,41 @@
 	var/whitelist_ckey
 	var/whitelist_charname
 
+/// Determines if the borg has the proper flags to show an overlay.
+/datum/robot_sprite/proc/sprite_flag_check(var/flag_to_check)
+	return (sprite_flags & flag_to_check)
+
 /datum/robot_sprite/proc/handle_extra_icon_updates(var/mob/living/silicon/robot/ourborg)
+	if(ourborg.resting) //Don't do ANY of the overlay code if we're resting. It just won't look right!
+		return
+	if(sprite_flag_check(ROBOT_HAS_SHIELD_SPEED_SPRITE))
+		if(ourborg.has_active_type(/obj/item/borg/combat/shield) && ourborg.has_active_type(/obj/item/borg/combat/mobility))
+			ourborg.add_overlay("[sprite_icon_state]-speed_shield")
+			return //Stop here. No need to add more overlays. Nothing else is compatible.
+
+	if(sprite_flag_check(ROBOT_HAS_SPEED_SPRITE) && ourborg.has_active_type(/obj/item/borg/combat/mobility))
+		ourborg.icon_state = "[sprite_icon_state]-roll"
+		return //Stop here. No need to add more overlays. Nothing else is compatible.
+
+	if(sprite_flag_check(ROBOT_HAS_SHIELD_SPRITE))
+		if(ourborg.has_active_type(/obj/item/borg/combat/shield))
+			var/obj/item/borg/combat/shield/shield = locate() in ourborg
+			if(shield && shield.active)
+				ourborg.add_overlay("[sprite_icon_state]-shield")
+
+	if(ourborg.has_active_type(/obj/item/gun/energy/robotic)) //We have a gun in our modules.
+		var/obj/item/gun/energy/robotic/gun = locate() in ourborg
+		for(gun)
+			if(!ourborg.has_active_type(gun)) //The gun we're checking is NOT selected. Pass over it.
+				continue
+			if(sprite_flag_check(ROBOT_HAS_GUN_SPRITE) && gun.gun_flag_check(COUNTS_AS_ROBOT_GUN))
+				ourborg.add_overlay("[sprite_icon_state]-gun")
+			if(sprite_flag_check(ROBOT_HAS_LASER_SPRITE) && gun.gun_flag_check(COUNTS_AS_ROBOT_LASER))
+				ourborg.add_overlay("[sprite_icon_state]-laser")
+			if(sprite_flag_check(ROBOT_HAS_TASER_SPRITE) && gun.gun_flag_check(COUNTS_AS_ROBOT_TASER))
+				ourborg.add_overlay("[sprite_icon_state]-taser")
+			if(sprite_flag_check(ROBOT_HAS_DISABLER_SPRITE) && gun.gun_flag_check(COUNTS_AS_ROBOT_DISABLER))
+				ourborg.add_overlay("[sprite_icon_state]-disabler")
```

Comment:
if(sprite_flag_check(ROBOT_HAS_BLADE_SPRITE) && gun.gun_flag_check(COUNTS_AS_ROBOT_BLADE))
				ourborg.add_overlay("[sprite_icon_state]-blade")

used for the /obj/item/melee/combat_borgblade